### PR TITLE
[Starrocks On ES] support sql explain for ES table

### PR DESCRIPTION
**Pull request by RowenWoo**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

```
 explain select * from es780 where login_time = 1653035909 and username = 'cccc';
```

result for `EsScanNode` snippet is:
```
.....
|   0:EsScanNode                                                                                                                                                                                                                                                                      |
|      TABLE: es780                                                                                                                                                                                                                                                                   |
|      PREDICATES: 1: login_time = 1653035909, 2: username = 'cccc'                                                                                                                                                                                                                   |
|      ES_QUERY_DSL: {"bool":{"must":[{"term":{"login_time":1653035909}},{"term":{"username":"cccc"}}]}} 
......
```

### CODE REVIEW

File: fe/fe-core/src/test/java/com/starrocks/external/elasticsearch/QueryTranslatorTest.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
+package com.starrocks.external.elasticsearch;
```
Comment (imay) [inline fe/fe-core/src/test/java/com/starrocks/external/elasticsearch/QueryTranslatorTest.java:1]: should add license header

File: fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryTranslator.java [hunk 2]
```diff
@@ -0,0 +1,224 @@
+// This file is made available under Elastic License 2.0.
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryTranslator.java:1]: If this is created by yourself, you need to use starrocks license header.

File: fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryTranslator.java [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static com.starrocks.external.elasticsearch.QueryBuilders.QueryBuilder;
+
+/**
+ * This class is responsible for translating the SR's conjuncts Expr into Elasticsearch's Query DSL
+ */
+public class QueryTranslator {
```
Comment (imay) [inline fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryTranslator.java:50]: I think you can implement this as a Visitor
And you implement them in the visit functions, such as `visitBinaryPredicate`

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 97 / 242 (40.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryConverter.java|0|97|00.00%|
|:large_blue_circle:|com/starrocks/planner/EsScanNode.java|0|14|00.00%|
|:large_blue_circle:|com/starrocks/common/io/FastByteArrayOutputStream.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryTranslator.java|75|102|73.53%|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryBuilders.java|20|26|76.92%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 94 / 157 (59.87%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/EsScanNode.java|0|22|00.00%|
|:large_blue_circle:|com/starrocks/common/io/FastByteArrayOutputStream.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryConverter.java|72|106|67.92%|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryBuilders.java|20|26|76.92%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 94 / 157 (59.87%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/EsScanNode.java|0|22|00.00%|
|:large_blue_circle:|com/starrocks/common/io/FastByteArrayOutputStream.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryConverter.java|72|106|67.92%|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryBuilders.java|20|26|76.92%|

File: fe/fe-core/src/main/java/com/starrocks/planner/EsScanNode.java [hunk 4]
```diff
@@ -237,23 +239,38 @@ protected String getNodeExplainString(String prefix, TExplainLevel detailLevel)
...
+        } else {
+            QueryConverter queryConverter = new QueryConverter();
+            QueryBuilders.QueryBuilder queryBuilder = queryConverter.convert(getConjuncts());
+            output.append(prefix).append("PREDICATES: ").append(
+                    getExplainString(conjuncts)).append("\n");
+            // reserved for later using: LOCAL_PREDICATES is processed by StarRocks EsScanNode
+            output.append(prefix).append("LOCAL_PREDICATES: ")
+                    .append(getExplainString(queryConverter.localConjuncts()))
+                    .append("\n");
+            output.append(prefix).append("REMOTE_PREDICATES: ")
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/planner/EsScanNode.java:255]: Shall we unify the naming style? Such as `Remote Predicates: `?

Comment (RowenWoo) [reply fe/fe-core/src/main/java/com/starrocks/planner/EsScanNode.java:255]: Are there any other places? I didn't notice this

File: fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryBuilders.java [hunk 5]
```diff
@@ -30,6 +38,39 @@
...
+         * Convert query to JSON format
+         *
+         * @param out used to generate JSON elements
+         * @throws IOException if IO error occurred
+         */
+        public abstract void toJson(JsonGenerator out) throws IOException;
+
+        @Override
+        public String toString() throws StarRocksESException {
+            FastByteArrayOutputStream out = new FastByteArrayOutputStream(256);
```
Comment (stephen-shelby) [inline fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryBuilders.java:56]: Is 256 a magic number?

File: fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryConverter.java [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+import com.starrocks.sql.ast.AstVisitor;
+import com.starrocks.thrift.TExprOpcode;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class QueryConverter extends AstVisitor<QueryBuilders.QueryBuilder, Void> {
+
+    // expr sets which can not be push down to Elasticsearch, SR BE should process
```
Comment (DorianZheng) [inline fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryConverter.java:30]: push->pushed

File: fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryBuilders.java (hunk 5 above)
Comment (RowenWoo) [reply fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryBuilders.java:56]: no, it's a initial size

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 92 / 153 (60.13%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/EsScanNode.java|0|22|00.00%|[242, 247, 248, 249, 250, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 265, 266, 267, 268, 270, 271, 272]|
|:large_blue_circle:|com/starrocks/common/io/FastByteArrayOutputStream.java|2|3|66.67%|[102]|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryConverter.java|72|106|67.92%|[37, 41, 49, 50, 52, 54, 57, 58, 59, 60, 61, 62, 63, 75, 83, 90, 93, 105, 106, 107, 124, 145, 146, 147, 181, 187, 191, 199, 210, 213, 217, 219, 221, 230]|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryBuilders.java|18|22|81.82%|[62, 63, 488, 489]|

**Review by DorianZheng — DISMISSED** [review]
LGTM

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 90 / 151 (59.60%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/planner/EsScanNode.java|0|22|00.00%|[242, 247, 248, 249, 250, 252, 253, 254, 255, 256, 257, 258, 259, 260, 261, 265, 266, 267, 268, 270, 271, 272]|
|:large_blue_circle:|com/starrocks/common/io/FastByteArrayOutputStream.java|2|3|66.67%|[102]|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryConverter.java|70|104|67.31%|[37, 41, 49, 50, 52, 54, 57, 58, 59, 60, 61, 62, 63, 75, 83, 90, 93, 105, 106, 107, 124, 145, 146, 147, 179, 185, 189, 197, 208, 211, 215, 217, 219, 228]|
|:large_blue_circle:|com/starrocks/external/elasticsearch/QueryBuilders.java|18|22|81.82%|[60, 61, 487, 488]|

File: fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryConverter.java [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+import com.starrocks.analysis.SlotRef;
+import com.starrocks.analysis.StringLiteral;
+import com.starrocks.sql.ast.AstVisitor;
+import com.starrocks.thrift.TExprOpcode;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class QueryConverter extends AstVisitor<QueryBuilders.QueryBuilder, Void> {
```
Comment (caneGuy) [inline fe/fe-core/src/main/java/com/starrocks/external/elasticsearch/QueryConverter.java:28]: name it as ConjunctConverter may be more concretely?

**Review by imay — APPROVED** [review]

**Pull request merged**